
![GiantHunter](logo.jpeg)


# Overview
The main function of GiantHunter is to identify NCLDV-like contigs from metagenomic data. The input of the program should be fasta files and the output will be a csv file showing the predictions. Since it is a deep learning model, if you have GPU units on your PC, we recommend that you use them to save time. 

If you have any trouble installing or using GiantHunter, you can post an issue or directly email us (jiayushang@cuhk.edu.hk). We welcome any suggestions.


### Quick install
*Note*: we suggest you to install all the package using conda (both miniconda and [Anaconda](https://anaconda.org/) are ok).

### Prepare the environment
1. When you use GiantHunter at the first time
```
git clone https://github.com/FuchuanQu/GiantHunter.git
cd GiantHunter/
conda env create -f GiantHunter.yaml -n gianthunter
conda activate gianthunter
pip install .

# run an example
gianthunter --contigs example/test.fasta --dbdir database --out test --threads 4
# In the test folder there should be predictions generated by GiantHunter.
```


2. If the example can be run without any bugs, you only need to activate your 'gianthunter' environment before using GiantHunter.
```
conda activate gianthunter
```


## Usage

```
gianthunter [--contigs INPUT_FA] [--out OUTPUT_CSV] [--reject THRESHOLD] [--midfolder DIR] [--threads NUM] [--dbdir DR] [--query_cover QC]
```

**Options**


      --contigs INPUT_FA
                            input fasta file
      --len MINIMUM_LEN
                            predict only for sequence >= len bp (default 3000)
      --proteins PROTEIN_FA
                            An optional protein file. If you have already annotated your contigs, you can use them as the inputs. 
                            Otherwise, GiantHunter will run prodigal to translate your contigs.
      --threads NUM
                            Number of threads to run GiantHunter (default maximum can be allocated)
      --dbdir DATABASE_DR
                            An optional path to store the database directory (default database/)
      --out OUTPUT_DR
                            The output csv file (prediction)
      --reject THRESHOLD
                            Threshold to reject contigs with a small fraction of proteins aligned. (default 0.3)
      --midfolder TEMP_DIR
                            Folder to store the intermediate files (default temp/)
      --query_cover QC
                            The QC value set for DIAMOND BLASTP, setting to 0 means no query-cover constrain (default 40) 

**Example**

Prediction on the example file:

    python run.py --contigs GiantHunter/example/test.fasta --out test/ --threads 8 --dbdir GiantHunter/database

Please always remember to input the correct path where you store the GiantHunter database.

### Outputs

There are four main outputs in the folder:

      GiantHunter_prediction.csv
                            The predictions for all the contigs
      
      giant_virus_contigs.fa
                            The detected Giant Virus in FASTA format
      
      giant_virus_proteins.fa
                            The predicted proteins of detected Giant Virus in FASTA format

      protein_annotation.csv
                            The protein annotation of the detected Giant Virus
      
    
    
### References

The paper is accepted by ISMB/ECCB 2025s. See you in Liverpool!

The arXiv version can be found via: [arXiv version](https://arxiv.org/abs/2501.15472)
